sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -88,7 +88,8 @@ class Analyzer(
     Batch("UDF", Once,
       HandleNullInputsForUDF),
     Batch("Cleanup", fixedPoint,
-      CleanupAliases)
+      CleanupAliases,
+      ResolveUpCast)
```

Comment:
I guess I can't find a case where this causes a problem, but its seems a little odd to have this here instead of as part of normal resolution.  I guess its fine because we are speculatively assuming that the UpCast is going to resolve into a cast that returns `dataType` anyway?

It seems slightly better to me to have `UpCast` be `unresolved` and fixed as we resolve the plan.  However, I would not block merging the PR on this.